File: gearted-backend/src/constants/airsoft-categories.ts
/**
 * Classification complète des catégories et sous-catégories de matériel Airsoft
 * Structure numérotée avec bullet points pour une navigation claire
 * Synchronisée avec la version mobile
 */

class AirsoftCategories {
  // 🪖 1. Répliques Airsoft
  static readonly REPLIQUES_AIRSOFT = '🪖 1. Répliques Airsoft';
  static readonly REPLIQUE_POING_GAZ = '• Répliques de poing - Gaz';
  static readonly REPLIQUE_POING_AEP = '• Répliques de poing - AEP';
  static readonly REPLIQUE_POING_CO2 = '• Répliques de poing - CO2';
  static readonly REPLIQUE_POING_SPRING = '• Répliques de poing - Spring';
  static readonly REPLIQUE_LONGUE_AEG = '• Répliques longues - AEG';
  static readonly REPLIQUE_LONGUE_AEG_BLOWBACK = '• Répliques longues - AEG Blowback';
  static readonly REPLIQUE_LONGUE_GBB = '• Répliques longues - GBB';
  static readonly REPLIQUE_LONGUE_SPRING = '• Répliques longues - Spring';

  // 🧤 2. Équipement de protection
  static readonly EQUIPEMENT_PROTECTION = '🧤 2. Équipement de protection';
  static readonly MASQUES = '• Masques';
  static readonly LUNETTES_PROTECTION = '• Lunettes de protection balistique';
  static readonly CASQUES = '• Casques';
  static readonly GANTS = '• Gants';
  static readonly PROTEGE_VISAGE = '• Protège-dents / Protège-visage';
  static readonly GENOUILLERES = '• Genouillères / Coudières';
  static readonly PORTE_PLAQUES = '• Porte-plaques / Porte-chargeurs';
  static readonly GILETS_TACTIQUES = '• Gilets tactiques / Chest Rigs';

  // 🎽 3. Tenues et camouflages
  static readonly TENUES_CAMOUFLAGES = '🎽 3. Tenues et camouflages';
  static readonly UNIFORMES = '• Uniformes (militaire, police, civil tactique)';
  static readonly CEINTURES_COMBAT = '• Ceintures de combat';
  static readonly CASQUETTES_CHAPEAUX = '• Casquettes / Chapeaux / Shemaghs';
  static readonly CHAUSSURES_RANGERS = '• Chaussures / Rangers';

  // 📦 4. Accessoires de réplique
  static readonly ACCESSOIRES_REPLIQUE = '📦 4. Accessoires de réplique';
  static readonly CHARGEURS = '• Chargeurs';
  static readonly SILENCIEUX_TRACERS = '• Silencieux / Tracers';
  static readonly ORGANES_VISEE = '• Organes de visée (Red Dot, lunettes, etc.)';
  static readonly RAILS_GRIPS = '• Rails & grips / poignées';
  static readonly LAMPES_TACTIQUES = '• Lampes tactiques / Lasers';
  static readonly GRENADES_AIRSOFT = '• Grenades airsoft (gaz ou ressort)';

  // ⚙️ 5. Pièces internes et upgrade
  static readonly PIECES_INTERNES_UPGRADE = '⚙️ 5. Pièces internes et upgrade';
  static readonly GEARBOX = '• Gearbox (mécanique AEG)';
  static readonly MOTEURS = '• Moteurs';
  static readonly PISTONS_RESSORTS = '• Pistons, ressorts, engrenages';
  static readonly CANON_PRECISION = '• Canon de précision';
  static readonly HOP_UP_JOINTS = '• Hop-up et joints';
  static readonly BLOC_DETENTE = '• Bloc détente';

  // 🛠 6. Outils et maintenance
  static readonly OUTILS_MAINTENANCE = '🛠 6. Outils et maintenance';
  static readonly KITS_NETTOYAGE = '• Kits de nettoyage';
  static readonly TOURNEVIS_CLEFS = '• Tournevis / Clés Allen';
  static readonly CHARGEURS_UNIVERSELS = '• Chargeurs universels';
  static readonly SAC_TRANSPORT = '• Sac de transport / mallette de réplique';

  // 📻 7. Communication & électronique
  static readonly COMMUNICATION_ELECTRONIQUE = '📻 7. Communication & électronique';
  static readonly TALKIES_WALKIES = '• Talkies-Walkies';
  static readonly PACKS_PTT = '• Packs PTT + casques';
  static readonly CAMERAS = '• Caméras (GoPro, tactiques)';
  static readonly UNITES_TRACAGE = '• Unités de traçage lumineuses (tracer units)';

  /**
   * Liste complète des catégories principales avec numérotation et emojis
   */
  static readonly MAIN_CATEGORIES: readonly string[] = [
    AirsoftCategories.REPLIQUES_AIRSOFT,
    AirsoftCategories.EQUIPEMENT_PROTECTION,
    AirsoftCategories.TENUES_CAMOUFLAGES,
    AirsoftCategories.ACCESSOIRES_REPLIQUE,
    AirsoftCategories.PIECES_INTERNES_UPGRADE,
    AirsoftCategories.OUTILS_MAINTENANCE,
    AirsoftCategories.COMMUNICATION_ELECTRONIQUE,
  ];

  /**
   * Liste complète de toutes les catégories et sous-catégories
   */
  static readonly ALL_CATEGORIES: readonly string[] = [
    // 🪖 1. Répliques Airsoft
    AirsoftCategories.REPLIQUES_AIRSOFT,
    AirsoftCategories.REPLIQUE_POING_GAZ,
    AirsoftCategories.REPLIQUE_POING_AEP,
    AirsoftCategories.REPLIQUE_POING_CO2,
    AirsoftCategories.REPLIQUE_POING_SPRING,
    AirsoftCategories.REPLIQUE_LONGUE_AEG,
    AirsoftCategories.REPLIQUE_LONGUE_AEG_BLOWBACK,
    AirsoftCategories.REPLIQUE_LONGUE_GBB,
    AirsoftCategories.REPLIQUE_LONGUE_SPRING,

    // 🧤 2. Équipement de protection
    AirsoftCategories.EQUIPEMENT_PROTECTION,
    AirsoftCategories.MASQUES,
    AirsoftCategories.LUNETTES_PROTECTION,
    AirsoftCategories.CASQUES,
    AirsoftCategories.GANTS,
    AirsoftCategories.PROTEGE_VISAGE,
    AirsoftCategories.GENOUILLERES,
    AirsoftCategories.PORTE_PLAQUES,
    AirsoftCategories.GILETS_TACTIQUES,

    // 🎽 3. Tenues et camouflages
    AirsoftCategories.TENUES_CAMOUFLAGES,
    AirsoftCategories.UNIFORMES,
    AirsoftCategories.CEINTURES_COMBAT,
    AirsoftCategories.CASQUETTES_CHAPEAUX,
    AirsoftCategories.CHAUSSURES_RANGERS,

    // 📦 4. Accessoires de réplique
    AirsoftCategories.ACCESSOIRES_REPLIQUE,
    AirsoftCategories.CHARGEURS,
    AirsoftCategories.SILENCIEUX_TRACERS,
    AirsoftCategories.ORGANES_VISEE,
    AirsoftCategories.RAILS_GRIPS,
    AirsoftCategories.LAMPES_TACTIQUES,
    AirsoftCategories.GRENADES_AIRSOFT,

    // ⚙️ 5. Pièces internes et upgrade
    AirsoftCategories.PIECES_INTERNES_UPGRADE,
    AirsoftCategories.GEARBOX,
    AirsoftCategories.MOTEURS,
    AirsoftCategories.PISTONS_RESSORTS,
    AirsoftCategories.CANON_PRECISION,
    AirsoftCategories.HOP_UP_JOINTS,
    AirsoftCategories.BLOC_DETENTE,

    // 🛠 6. Outils et maintenance
    AirsoftCategories.OUTILS_MAINTENANCE,
    AirsoftCategories.KITS_NETTOYAGE,
    AirsoftCategories.TOURNEVIS_CLEFS,
    AirsoftCategories.CHARGEURS_UNIVERSELS,
    AirsoftCategories.SAC_TRANSPORT,

    // 📻 7. Communication & électronique
    AirsoftCategories.COMMUNICATION_ELECTRONIQUE,
    AirsoftCategories.TALKIES_WALKIES,
    AirsoftCategories.PACKS_PTT,
    AirsoftCategories.CAMERAS,
    AirsoftCategories.UNITES_TRACAGE,
  ];

  /**
   * Mapping des catégories principales vers leurs sous-catégories
   */
  static readonly CATEGORY_MAPPING: Record<string, readonly string[]> = {
    [AirsoftCategories.REPLIQUES_AIRSOFT]: [
      AirsoftCategories.REPLIQUE_POING_GAZ,
      AirsoftCategories.REPLIQUE_POING_AEP,
      AirsoftCategories.REPLIQUE_POING_CO2,
      AirsoftCategories.REPLIQUE_POING_SPRING,
      AirsoftCategories.REPLIQUE_LONGUE_AEG,
      AirsoftCategories.REPLIQUE_LONGUE_AEG_BLOWBACK,
      AirsoftCategories.REPLIQUE_LONGUE_GBB,
      AirsoftCategories.REPLIQUE_LONGUE_SPRING,
    ],
    [AirsoftCategories.EQUIPEMENT_PROTECTION]: [
      AirsoftCategories.MASQUES,
      AirsoftCategories.LUNETTES_PROTECTION,
      AirsoftCategories.CASQUES,
      AirsoftCategories.GANTS,
      AirsoftCategories.PROTEGE_VISAGE,
      AirsoftCategories.GENOUILLERES,
      AirsoftCategories.PORTE_PLAQUES,
      AirsoftCategories.GILETS_TACTIQUES,
    ],
    [AirsoftCategories.TENUES_CAMOUFLAGES]: [
      AirsoftCategories.UNIFORMES,
      AirsoftCategories.CEINTURES_COMBAT,
      AirsoftCategories.CASQUETTES_CHAPEAUX,
      AirsoftCategories.CHAUSSURES_RANGERS,
    ],
    [AirsoftCategories.ACCESSOIRES_REPLIQUE]: [
      AirsoftCategories.CHARGEURS,
      AirsoftCategories.SILENCIEUX_TRACERS,
      AirsoftCategories.ORGANES_VISEE,
      AirsoftCategories.RAILS_GRIPS,
      AirsoftCategories.LAMPES_TACTIQUES,
      AirsoftCategories.GRENADES_AIRSOFT,
    ],
    [AirsoftCategories.PIECES_INTERNES_UPGRADE]: [
      AirsoftCategories.GEARBOX,
      AirsoftCategories.MOTEURS,
      AirsoftCategories.PISTONS_RESSORTS,
      AirsoftCategories.CANON_PRECISION,
      AirsoftCategories.HOP_UP_JOINTS,
      AirsoftCategories.BLOC_DETENTE,
    ],
    [AirsoftCategories.OUTILS_MAINTENANCE]: [
      AirsoftCategories.KITS_NETTOYAGE,
      AirsoftCategories.TOURNEVIS_CLEFS,
      AirsoftCategories.CHARGEURS_UNIVERSELS,
      AirsoftCategories.SAC_TRANSPORT,
    ],
    [AirsoftCategories.COMMUNICATION_ELECTRONIQUE]: [
      AirsoftCategories.TALKIES_WALKIES,
      AirsoftCategories.PACKS_PTT,
      AirsoftCategories.CAMERAS,
      AirsoftCategories.UNITES_TRACAGE,
    ],
  };

  /**
   * Obtenir la catégorie principale d'une sous-catégorie
   */
  static getMainCategory(subCategory: string): string | null {
    for (const [mainCategory, subCategories] of Object.entries(this.CATEGORY_MAPPING)) {
      if (subCategories.includes(subCategory)) {
        return mainCategory;
      }
    }
    return null;
  }

  /**
   * Obtenir les sous-catégories d'une catégorie principale
   */
  static getSubCategories(mainCategory: string): readonly string[] {
    return this.CATEGORY_MAPPING[mainCategory] || [];
  }

  /**
   * Vérifier si une catégorie est valide
   */
  static isValidCategory(category: string): boolean {
    return this.ALL_CATEGORIES.includes(category);
  }

  /**
   * Vérifier si une catégorie est une catégorie principale
   */
  static isMainCategory(category: string): boolean {
    return this.MAIN_CATEGORIES.includes(category);
  }

  /**
   * Obtenir toutes les catégories dans une structure hiérarchique
   */
  static getCategoryHierarchy(): Record<string, readonly string[]> {
    return this.CATEGORY_MAPPING;
  }
}

export default AirsoftCategories;

File: gearted-mobile/lib/core/constants/airsoft_categories.dart
/// Classification complète des catégories et sous-catégories de matériel Airsoft
/// Structure numérotée avec bullet points pour une navigation claire
/// Synchronisée avec la version backend

class AirsoftCategories {
  // 🪖 1. Répliques Airsoft
  static const String repliquesAirsoft = '🪖 1. Répliques Airsoft';
  static const String repliquePoingGaz = '• Répliques de poing - Gaz';
  static const String repliquePoingAEP = '• Répliques de poing - AEP';
  static const String repliquePoingCO2 = '• Répliques de poing - CO2';
  static const String repliquePoingSpring = '• Répliques de poing - Spring';
  static const String repliqueLongueAEG = '• Répliques longues - AEG';
  static const String repliqueLongueAEGBlowback = '• Répliques longues - AEG Blowback';
  static const String repliqueLongueGBB = '• Répliques longues - GBB';
  static const String repliqueLongueSpring = '• Répliques longues - Spring';

  // 🧤 2. Équipement de protection
  static const String equipementProtection = '🧤 2. Équipement de protection';
  static const String masques = '• Masques';
  static const String lunettesProtection = '• Lunettes de protection balistique';
  static const String casques = '• Casques';
  static const String gants = '• Gants';
  static const String protegeVisage = '• Protège-dents / Protège-visage';
  static const String genouilleres = '• Genouillères / Coudières';
  static const String portePlaques = '• Porte-plaques / Porte-chargeurs';
  static const String giletsTactiques = '• Gilets tactiques / Chest Rigs';

  // 🎽 3. Tenues et camouflages
  static const String tenuesCamouflages = '🎽 3. Tenues et camouflages';
  static const String uniformes = '• Uniformes (militaire, police, civil tactique)';
  static const String ceinturesCombat = '• Ceintures de combat';
  static const String casquettesChapeaux = '• Casquettes / Chapeaux / Shemaghs';
  static const String chaussuresRangers = '• Chaussures / Rangers';

  // 📦 4. Accessoires de réplique
  static const String accessoiresReplique = '📦 4. Accessoires de réplique';
  static const String chargeurs = '• Chargeurs';
  static const String silencieuxTracers = '• Silencieux / Tracers';
  static const String organesVisee = '• Organes de visée (Red Dot, lunettes, etc.)';
  static const String railsGrips = '• Rails & grips / poignées';
  static const String lampesTactiques = '• Lampes tactiques / Lasers';
  static const String grenadesAirsoft = '• Grenades airsoft (gaz ou ressort)';

  // ⚙️ 5. Pièces internes et upgrade
  static const String piecesInternesUpgrade = '⚙️ 5. Pièces internes et upgrade';
  static const String gearbox = '• Gearbox (mécanique AEG)';
  static const String moteurs = '• Moteurs';
  static const String pistonsRessorts = '• Pistons, ressorts, engrenages';
  static const String canonPrecision = '• Canon de précision';
  static const String hopUpJoints = '• Hop-up et joints';
  static const String blocDetente = '• Bloc détente';

  // 🛠 6. Outils et maintenance
  static const String outilsMaintenance = '🛠 6. Outils et maintenance';
  static const String kitsNettoyage = '• Kits de nettoyage';
  static const String tournevisClefs = '• Tournevis / Clés Allen';
  static const String chargeursUniversels = '• Chargeurs universels';
  static const String sacTransport = '• Sac de transport / mallette de réplique';

  // 📻 7. Communication & électronique
  static const String communicationElectronique = '📻 7. Communication & électronique';
  static const String talkiesWalkies = '• Talkies-Walkies';
  static const String packsPTT = '• Packs PTT + casques';
  static const String cameras = '• Caméras (GoPro, tactiques)';
  static const String unitesTracage = '• Unités de traçage lumineuses (tracer units)';

  /// Liste complète des catégories principales avec numérotation et emojis
  static const List<String> mainCategories = [
    repliquesAirsoft,
    equipementProtection,
    tenuesCamouflages,
    accessoiresReplique,
    piecesInternesUpgrade,
    outilsMaintenance,
    communicationElectronique,
  ];

  /// Liste complète de toutes les catégories et sous-catégories
  static const List<String> allCategories = [
    // 🪖 1. Répliques Airsoft
    repliquesAirsoft,
    repliquePoingGaz,
    repliquePoingAEP,
    repliquePoingCO2,
    repliquePoingSpring,
    repliqueLongueAEG,
    repliqueLongueAEGBlowback,
    repliqueLongueGBB,
    repliqueLongueSpring,

    // 🧤 2. Équipement de protection
    equipementProtection,
    masques,
    lunettesProtection,
    casques,
    gants,
    protegeVisage,
    genouilleres,
    portePlaques,
    giletsTactiques,

    // 🎽 3. Tenues et camouflages
    tenuesCamouflages,
    uniformes,
    ceinturesCombat,
    casquettesChapeaux,
    chaussuresRangers,

    // 📦 4. Accessoires de réplique
    accessoiresReplique,
    chargeurs,
    silencieuxTracers,
    organesVisee,
    railsGrips,
    lampesTactiques,
    grenadesAirsoft,

    // ⚙️ 5. Pièces internes et upgrade
    piecesInternesUpgrade,
    gearbox,
    moteurs,
    pistonsRessorts,
    canonPrecision,
    hopUpJoints,
    blocDetente,

    // 🛠 6. Outils et maintenance
    outilsMaintenance,
    kitsNettoyage,
    tournevisClefs,
    chargeursUniversels,
    sacTransport,

    // 📻 7. Communication & électronique
    communicationElectronique,
    talkiesWalkies,
    packsPTT,
    cameras,
    unitesTracage,
  ];

  /// Mapping des catégories principales vers leurs sous-catégories
  static const Map<String, List<String>> categoryMapping = {
    repliquesAirsoft: [
      repliquePoingGaz,
      repliquePoingAEP,
      repliquePoingCO2,
      repliquePoingSpring,
      repliqueLongueAEG,
      repliqueLongueAEGBlowback,
      repliqueLongueGBB,
      repliqueLongueSpring,
    ],
    equipementProtection: [
      masques,
      lunettesProtection,
      casques,
      gants,
      protegeVisage,
      genouilleres,
      portePlaques,
      giletsTactiques,
    ],
    tenuesCamouflages: [
      uniformes,
      ceinturesCombat,
      casquettesChapeaux,
      chaussuresRangers,
    ],
    accessoiresReplique: [
      chargeurs,
      silencieuxTracers,
      organesVisee,
      railsGrips,
      lampesTactiques,
      grenadesAirsoft,
    ],
    piecesInternesUpgrade: [
      gearbox,
      moteurs,
      pistonsRessorts,
      canonPrecision,
      hopUpJoints,
      blocDetente,
    ],
    outilsMaintenance: [
      kitsNettoyage,
      tournevisClefs,
      chargeursUniversels,
      sacTransport,
    ],
    communicationElectronique: [
      talkiesWalkies,
      packsPTT,
      cameras,
      unitesTracage,
    ],
  };

  /// Obtenir les icônes correspondantes aux catégories principales
  static const Map<String, int> categoryIcons = {
    repliquesAirsoft: 0xe52f, // Icons.sports_motorsports
    equipementProtection: 0xe3bb, // Icons.security
    tenuesCamouflages: 0xe3c6, // Icons.checkroom
    accessoiresReplique: 0xe5f1, // Icons.build
    piecesInternesUpgrade: 0xe5ca, // Icons.precision_manufacturing
    outilsMaintenance: 0xe5f5, // Icons.build_circle
    communicationElectronique: 0xe324, // Icons.radio
  };

  /// Obtenir les sous-catégories d'une catégorie principale
  static List<String> getSubCategories(String mainCategory) {
    return categoryMapping[mainCategory] ?? [];
  }

  /// Vérifier si une catégorie est une catégorie principale
  static bool isMainCategory(String category) {
    return mainCategories.contains(category);
  }

  /// Obtenir la catégorie principale d'une sous-catégorie
  static String? getMainCategory(String subCategory) {
    for (final entry in categoryMapping.entries) {
      if (entry.value.contains(subCategory)) {
        return entry.key;
      }
    }
    return null;
  }

  /// Helper functions to clean category names for display
  
  /// Remove emoji and number from category name for clean display
  static String getCleanCategoryName(String category) {
    // Remove emoji and number pattern like "🪖 1. " or "• "
    return category.replaceAll(RegExp(r'^[🪖🧤🎽📦⚙️🛠📻]\s*\d+\.\s*'), '')
                  .replaceAll(RegExp(r'^•\s*'), '');
  }
  
  /// Get clean main category names without emojis and numbers
  static List<String> get cleanMainCategories {
    return mainCategories.map((category) => getCleanCategoryName(category)).toList();
  }
  
  /// Get clean subcategories for a main category
  static List<String> getCleanSubCategories(String mainCategory) {
    final subCategories = getSubCategories(mainCategory);
    return subCategories.map((subCategory) => getCleanCategoryName(subCategory)).toList();
  }
  
  /// Get all clean subcategories (without main categories)
  static List<String> get allCleanSubCategories {
    List<String> allSubs = [];
    for (String mainCategory in mainCategories) {
      allSubs.addAll(getCleanSubCategories(mainCategory));
    }
    return allSubs;
  }
}
